### PR TITLE
add ftp storage, add getFileMeta, setFileMeta functions

### PR DESCRIPTION
**Pull request by egorpro123**

### CODE REVIEW

File: storages/ftp.js [hunk 1]
```diff
@@ -0,0 +1,133 @@
+/* eslint-disable max-len */
```
Comment (vedi) [inline storages/ftp.js]: remove it

File: storages/ftp.js [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+client = Bb.promisifyAll(client);
+
+const utils = require('../lib/utils');
+
+const requireOptions = utils.requireOptions;
+
+
+/**
+ *
+ * options:
```
Comment (vedi) [inline storages/ftp.js:22]: it does not look like JSDoc. What a standard is it?

File: storages/ftp.js [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+      client.once('ready', () => resolve(serverMessage));
+      client.once('error', err => reject(err));
+      client.connect(this.options);
+    });
+  },
+
+  getStream: function (fileMeta) {
+    const vals = {};
+
+    let filename = fileMeta.filename;
```
Comment (vedi) [inline storages/ftp.js]: camelCase it

File: storages/ftp.js [hunk 4]
```diff
@@ -0,0 +1,133 @@
...
+          stream,
+        };
+      }).catch(function(e) {
+        client.end();
+        throw e;
+      });
+  },
+
+  _fullRemoteFileName(filename) {
+    if (this.options.uploadPath) return path.join(this.options.uploadPath, fileName);
```
Comment (vedi) [inline storages/ftp.js]: always add brackets `{}`, please

File: storages/ftp.js [hunk 5]
```diff
@@ -0,0 +1,133 @@
...
+        return this.deleteFile(prevRemotePath);
+    }).then(() => {
+      return this.putFile(fromPath, options);
+    })
+  },
+
+  deleteFile: function (fileMeta) {
+    this._connect()
+      .then(function (msg) {
+        return new Bb((resolve, reject) => {
```
Comment (vedi) [inline storages/ftp.js]: `Bb.fromCallback` will look better

**Review by vedi — CHANGES_REQUESTED** [review]

File: storages/ftp.js (hunk 1 above)
Comment (egorpro123) [reply storages/ftp.js]: ok

File: storages/ftp.js (hunk 2 above)
Comment (egorpro123) [reply storages/ftp.js:22]: Its just comment. Converted to JSDoc.

File: storages/ftp.js (hunk 3 above)
Comment (egorpro123) [reply storages/ftp.js]: fixed

File: storages/ftp.js (hunk 4 above)
Comment (egorpro123) [reply storages/ftp.js]: No problem, fixed

File: storages/ftp.js (hunk 5 above)
Comment (egorpro123) [reply storages/ftp.js]: fixed

File: storages/ftp.js [hunk 6]
```diff
@@ -0,0 +1,140 @@
...
+      .try(() => {
+        if (prevRemotePath !== newRemotePath) {
+          return this.deleteFile(prevRemotePath);
+        }
+      })
+      .then(() => this.putFile(fromPath, options));
+  }
+
+  deleteFile(fileMeta) {
+    return this._connect()
```
Comment (vedi) [inline storages/ftp.js]:
```
return this
  ._connect()
  .then(() => Bb.fromCallback((callback) => {
    client.delete(this._fullRemoteFileName(fileMeta.fileName), callback);
  })
  .finally(() => {
    client.end();
  });
```

**Review by vedi — CHANGES_REQUESTED** [review]

File: storages/ftp.js (hunk 6 above)
Comment (egorpro123) [reply storages/ftp.js]: Thanks. I will correct

File: storages/ftp.js [hunk 7]
```diff
@@ -0,0 +1,143 @@
+/**
+ * Created by vedi on 14/04/17.
+ */
+
+'use strict';
+
+const Bb = require('bluebird');
+const Client = require('ftp');
```
Comment (vedi) [inline storages/ftp.js:8]: no ftp in dependencies

**Review by vedi — CHANGES_REQUESTED** [review]

**Comment by vedi** [comment]
pushed to `dev`

**Pull request closed without merging**